### PR TITLE
Login

### PR DESCRIPTION
**Pull request by Hepic**

### CODE REVIEW

**Comment by dionyziz** [comment]
Fix build

**Comment by dionyziz** [comment]
Screenshot

**Comment by Hepic** [comment]
pic3
pic2

**Comment by Hepic** [comment]
pic

**Comment by dionyziz** [comment]
rebase

**Comment by dionyziz** [comment]
Screenshot for "Το ψευδώνυμο δεν υπάρχει." case

File: client/locales/el.json [hunk 1]
```diff
@@ -6,7 +6,15 @@
             "empty": "Γράψε ένα ψευδώνυμο.",
             "length": "Το ψευδώνυμο πρέπει να είναι έως 20 γράμματα.",
             "chars": "Το ψευδώνυμο πρέπει να περιλαμβάνει μόνο γράμματα ή αριθμούς.",
-            "taken": "Το ψευδώνυμο το έχει άλλος."
+            "taken": "Το ψευδώνυμο το έχει άλλος.",
+            "free": "Το ψευδώνυμο δεν υπάρχει. [Χρησιμοποίησέ το]"
```
Comment (dionyziz) [inline client/locales/el.json]: This needs to be a link, but we can take care of it later.

File: client/login.jsx [hunk 2]
```diff
@@ -76,9 +119,24 @@ var LoginForm = React.createClass({
...
+            'hidden': this.state.validationStateUsername === true && this.state.validationStatePassword === true
         });
 
+        var haveAccLink = null,
+            passwordField = null;
+
+        if (this.state.type == 'simpleLogin') {
+            haveAccLink = <div onClick={this.handleClick}>Έχω ήδη ψευδώνυμο</div>;
+        }
+
```
Comment (dionyziz) [inline client/login.jsx]: avoid blank lines between if/else

File: client/login.jsx [hunk 3]
```diff
@@ -22,49 +26,88 @@ var LoginForm = React.createClass({
...
         }
 
-        this._handleError(validationState);
+        this._handleError(validationStateUsername, this.state.validationStatePassword);
     },
-    onError(error) {
-        this._handleError(error);
+    _onPasswordChanged(password) {
+        var validationStatePassword = this._validatePassword(password);
+        this.setState({validationStatePassword: validationStatePassword});
```
Comment (dionyziz) [inline client/login.jsx]: better written as:

```
this.setState({validationStatePassword});
```

File: client/login.jsx [hunk 4]
```diff
@@ -22,49 +26,88 @@ var LoginForm = React.createClass({
...
+        else if (validationStatePassword !== true) {
+            this.setState({
+                errorStr: i18n.t('passwordSet.errors.' + validationStatePassword)
+            });
+        }
     },
     _onUsernameChanged(username) {
-        var validationState = this._validate(username);
+        var validationStateUsername = this._validateUsername(username);
+        this.setState({validationStateUsername: validationStateUsername});
```
Comment (dionyziz) [inline client/login.jsx]: better written as:

```
this.setState({validationStateUsername});
```

File: client/login.jsx [hunk 5]
```diff
@@ -22,49 +26,88 @@ var LoginForm = React.createClass({
...
         }
 
-        if (this.state.validationState !== true) {
+        if (this.state.type == 'passwordLogin' && !this.state.passwordChanged) {
+            this._onPasswordChanged('');
+            return;
+        }
+
+        if ((this.state.validationStateUsername !== true) ||
+           (this.state.type == 'passwordLogin' && this.state.validationStatePassword !== true)) {
```
Comment (dionyziz) [inline client/login.jsx]: For readability, this is better split into a separate if: First, check if the username validation failed and return. Next, check if the password validation failed and return:

```
if (this.state.validationStateUsername !== true) {
    return;
}
if (this.state.type == 'passwordLogin') {
    if (this.satte.validationStatePassword !== true) {
        return;
    }
}
```

File: client/login.jsx [hunk 6]
```diff
@@ -22,49 +26,88 @@ var LoginForm = React.createClass({
...
             return;
         }
 
-        this.props.onLoginIntention(this.state.username);
+        this.props.onLoginIntention(this.state.username, this.state.password);
+    },
+    handleClick() {
```
Comment (dionyziz) [inline client/login.jsx]: This is a bad event name for what it does. Perhaps `switchToPasswordLogin`?

File: client/ting.jsx [hunk 7]
```diff
@@ -43,9 +44,9 @@ const Ting = React.createClass({
         const URL = window.location.hostname + ':' + config.port;
         this._socket = io.connect(URL, {secure: config.websocket.secure});
 
-        this._socket.on('login-response', ({success, people, error}) => {
+        this._socket.on('login-response', ({success, people, errorUsername, errorPassword}) => {
```
Comment (dionyziz) [inline client/ting.jsx]: This requires arch changes. Please make a separate PR. I recommend we keep the `error` parameter and have various errors there in this form:

username-empty
username-length
username-taken
username-chars
password-empty
password-incorrect

File: client/login.jsx [hunk 8]
```diff
@@ -5,66 +5,125 @@ var classNames = require('classnames');
...
+        else if (this.state.validationStatePassword !== true) {
+            this._handleError(this.state.validationStatePassword);
+        }
+        else {
+            this._handleError(true);
+        }
+    },
+    _onPasswordChanged(password) {
+        var validationStatePassword = this._validatePassword(password);
+        this.setState({validationStatePassword});
```
Comment (gtklocker) [inline client/login.jsx]: These can be merged into one state change.

**Comment by gtklocker** [comment]
This looks like it's on the right path. Please fix the eslint issues so that it builds and we can reevaluate it afterwards.

File: client/index.css [hunk 9]
```diff
@@ -105,23 +105,55 @@
...
+    right: 3px;
+    border-left: 10px solid transparent;
+    border-right: 10px solid transparent;
+    border-bottom: 10px solid LightSlateGray;
+}
+.top .menu li {
+    margin: 15px 10px;
+    color: white;
+    cursor: pointer;
+}
```
Comment (Hepic) [inline client/index.css]: a

File: client/login.jsx (hunk 8 above)
Comment (Hepic) [inline client/login.jsx]: What do you mean by "these" ?

Comment (gtklocker) [inline client/login.jsx]: Lines 69 and 71.

**Comment by dionyziz** [comment]
Blocked by #152

**Comment by dionyziz** [comment]
Fix build

File: client/ting.jsx [hunk 10]
```diff
@@ -59,32 +65,43 @@ const Ting = React.createClass({
         });
 
         this._socket.on('message', (data) => {
-            this.refs.history.onMessage(data);
+            if (!this.state.active && data.username != this.state.intendedUsername) {
+                this.setState({
```
Comment (Hepic) [inline client/ting.jsx]: So do you suggest to add every 'state' variable into redux store even if that variable is used only from one specific file ?

File: client/avatar.jsx [hunk 11]
```diff
@@ -5,10 +5,17 @@ var Avatar = React.createClass({
...
+
+        if(this.props.username != null) {
+            var src = this._getAvatar(this.props.username);
+            ret = <img src={src} alt={this.props.username} className='avatar' onClick={this.props.toggleUserMenu} />;
+        }
 
         return (
-            <img src={src} alt={this.props.username} className='avatar' />
+            <div>
+                {ret}
```
Comment (Hepic) [inline client/avatar.jsx]: Yes, because Avatar react class is used in the topbar. So until the user login, the 'username' is empty. I need to return something each time so I return an empty div.

**Comment by dionyziz** [comment]
Superseded by #209. Redux will be added later.
